Made edit metadata tab fields dynamic and added entity type support

alexandrevryghem opened the pull request:
## References
* Related #2653

## Description
Refactored the `DsoEditMetadataValueComponent` to make it easier to customize the edit behaviour of certain fields. By default there is a distinction between 3 types of fields:
- `DsoEditMetadataTextFieldComponent`: Regular free text fields
- `DsoEditMetadataAuthorityFieldComponent`: For authority controlled fields (this contains all the logic from #2653)
- `DsoEditMetadataEntityFieldComponent`: For the `dspace.entity.type` field, this will ensure that no invalid entity types are submitted
  <img width="1129" alt="image" src="https://github.com/user-attachments/assets/70ada3ab-3314-44d8-80da-7f519431e07c">

This separation will make it easy to introduce new types in the future like for example one for dates to ensure that they always follow the [ISO 8601 standard](https://www.iso.org/iso-8601-date-and-time-format.html). 

## Instructions for Reviewers
List of changes in this PR:
* `DsoEditMetadataValueFieldLoaderComponent`: This is the dynamic component loader that is used to load one of the edit metadata value components (textarea/input/select)
* `DsoEditMetadataTextFieldComponent`: The regular text area field
* `DsoEditMetadataAuthorityFieldComponent`: Extracted all the code from #2653 into this new component, because it started to become too big to stay in the same file
* `DsoEditMetadataEntityFieldComponent`: This is the field that contains a dropdown field with all the values that can be filled in for `dspace.entity.type`. This field is similar to the one in the edit/create collection form.
* `Context`: 2 new contexts have been created to make it possible to differentiate between editing an existing field and adding a new field. Currently the same components are used for both

**Guidance for how to test or review this PR:**
- Verify that you can still submit values for regular text fields
- Verify that you can still submit values for authority controlled fields
- Verify that `dspace.entity.type` fields forces you to select one of the configured types
- Verify that the `dspace.entity.type` field also works when entities are disabled
- Test it both for creating new fields and editing existing ones

## Checklist
- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented:
@alexandrevryghem : Can you rebase this against `main`?  It's had merge conflicts for some time, and it'll be easier to review/test if those are resolved.  Thanks!

alexandrevryghem commented:
@tdonohue: I fixed the merge conflicts. I did have to manually port the changes from #3608 so it's probably also best to test that functionality again as well. I also replaced the custom dropdown icons with the built-in ones from bootstrap (see https://github.com/DSpace/dspace-angular/pull/3722/commits/882fe66cb38d42a273b15bddd9c445139673c565). I did this because for some reason the dropdown icons were being rendered as triangles again instead of arrows
<img width="1310" alt="image" src="https://github.com/user-attachments/assets/f5091421-c2e2-401d-ad91-a02e86db9eef" />
<img width="1310" alt="image" src="https://github.com/user-attachments/assets/07912347-57e0-4605-8b73-388b57035494" />

tdonohue commented:
NOTE: This PR has merge conflicts caused by the merger of #3997.  You will need to rebase this PR on the latest `main` and update it to use ["control flow" syntax](https://angular.dev/guide/templates/control-flow) by replacing any `ngIf`, `ngFor`, or `ngSwitch` usages in HTML templates with `@if`, `@for` or `@switch`.  See the control flow docs or the above PR for more details.

tdonohue commented on src/app/dso-shared/dso-edit-metadata/dso-edit-metadata-value-field/dso-edit-metadata-field.service.ts:
```diff
@@ -0,0 +1,52 @@
+@Injectable({
+  providedIn: 'root',
+})
+export class DsoEditMetadataFieldService {
```
Please add a basic description for this service in a code comment / TypeDocs

tdonohue commented on src/app/dso-shared/dso-edit-metadata/dso-edit-metadata-value-field/dso-edit-metadata-value-field-loader/dso-edit-metadata-value-field-loader.component.ts:
```diff
@@ -0,0 +1,84 @@
+    DynamicComponentLoaderDirective,
+  ],
+})
+export class DsoEditMetadataValueFieldLoaderComponent extends AbstractComponentLoaderComponent<Component> {
```
Again, it'd be good to describe this component class with a code comment

tdonohue commented on src/app/dso-shared/dso-edit-metadata/dso-edit-metadata-value-field/dso-edit-metadata-value-field-loader/dso-edit-metadata-value-field.decorator.ts:
```diff
@@ -0,0 +1,59 @@
+ * @param context The context to match
+ * @param theme the theme to match
+ */
+export function getDsoEditMetadataValueFieldComponent(type: EditMetadataValueFieldType, context: Context = DEFAULT_CONTEXT, theme = DEFAULT_THEME) {
```
Would it be possible to add basic specs for this function?  Codecov is warning that every line of this function appears to not be covered by tests.   I do realize though that this function is not too complex, but it might be nice to have tests nonetheless.  (But if it's too difficult / time consuming, we can skip it)

alexandrevryghem replied:
I started creating some custom tests for this, but I don't think we should include those, because if people want to customize some components and extend the `map` with their own components, the tests would also fail. To prevent this we would need to override the map in the tests, but this would require us from switching that variable from a `const` to a `let`, and I'm not sure that would be a good idea

tdonohue replied:
I see your point.  We can keep this as-is then.  As I noted initially, it was more of a "nice to have" thing, if it didn't cause problems.

tdonohue reviewed (changes requested):
@alexandrevryghem : Thanks for this PR!  I gave it a code review today and it looks good overall, but I have a few minor requests (mostly for code comments).

I also tested the Edit Item -> Edit Metadata page pretty throughly, trying authority backed fields (like dc.type, dc.publisher, dc.contributor.author), hierarchical fields (like dc.subject), the dspace.entity.type field and normal plain text fields.  Everything appears to still be working properly, as I'm seeing the proper dropdowns or hierarchical popups based on the configured fields.  So, I couldn't find any object bugs with your newly refactored code & I like the new behavior of `dspace.entity.type`.  Thanks again!

alexandrevryghem commented on src/app/dso-shared/dso-edit-metadata/dso-edit-metadata-value-field/abstract-dso-edit-metadata-value-field.component.ts:
```diff
@@ -0,0 +1,54 @@
+  selector: 'ds-dso-edit-metadata-entity-field',
+  template: '',
+})
+export abstract class AbstractDsoEditMetadataValueFieldComponent {
```
Selectors are required for every `@Component` decorator, and we can't remove the `@Component` all together because the class uses `@Input`s & `@Output`s which require the `@Component` decorator on their class. I do agree though that the selector should be named differently, my bad 🫣

tdonohue reviewed (approved):
👍 Thanks @alexandrevryghem for the quick updates!  This looks good to me now. My prior feedback has been addressed & as I said previously I couldn't find any bugs after testing all types of fields.

I'll wait for @atarix83 to give this a second set of eyes (as it's a larger PR), but I feel this looks good enough to merge.

tdonohue commented:
As agreed to in yesterday's DevMtg, I'm merging this with +1 votes.  @atarix83 noted that he has no objections to this work, and no one else objected to merging this for 9.0.  Thanks again @alexandrevryghem !
